feat(date-picker): added date picker component

marijohannessen opened the pull request:
## Overview

Resolves https://github.ibm.com/Bluemix/design-system-website/issues/661

Added Date Picker component.

The component is using [flatpickr](https://chmln.github.io/flatpickr/) as its base for the calendar. There are 3 different variations: `no-calendar`, `basic` and `range`.

**Visual review:** http://date-picker-example.stage1.mybluemix.net/components/date-picker/

## Testing / Reviewing

Do `npm install` before reviewing.

marijohannessen committed:
feat(date-picker): initial work

marijohannessen committed:
feat(date-picker): added more examples

marijohannessen committed:
feat(date-picker): added test file

marijohannessen committed:
feat(date-picker): temp removed a11y from testing

asudoh commented on src/components/date-picker/date-picker.js:
```diff
@@ -0,0 +1,227 @@
+    const day = date.getDate() < 10
+    ? `0${date.getDate()}`
+    : date.getDate();
+    return `${month} / ${day} / ${date.getFullYear()}`;
```
Different countries have different date format. I’d suggest using [`Intl.DateTimeFormat`](https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Global_Objects/DateTimeFormat#Examples).

marijohannessen replied:
Hi @asudoh , I changed the `formatDate` function a bit, do you think this is good enough or do we need to add more logic to this?

marijohannessen committed:
feat(date-picker): added back test

asudoh commented on src/components/date-picker/date-picker.js:
```diff
@@ -0,0 +1,227 @@
+    const date = this.element.querySelector(this.options.selectorDatePickerInput);
+    const calendar = new Flatpickr(date, {
+      allowInput: true,
+      dateFormat: 'm/d/Y',
```
It’s a bit of repeating, but, for example, “2017/5/31” is used in Japan instead of “5/31/2017”. Seems that Flatpickr has [custom `formatDate()` config](https://github.com/chmln/flatpickr/commit/a1c3548b7046d07c4d10e0ae8efdb01442337587) as well ass [custom `parseDate()` config](https://github.com/chmln/flatpickr/blob/a1c3548b7046d07c4d10e0ae8efdb01442337587/src/flatpickr.js#L1103), and thus I’d suggest using those - The implementation should use `Into.DateTimeFormat`.

hellobrian commented on package.json:
```diff
@@ -30,9 +30,11 @@
     "lodash.debounce": "^4.0.8"
   },
   "devDependencies": {
+    "@ibma/karma-ibma": "^2.2.0",
```
It looks we removed this from our package.json in [`master` branch](https://github.com/carbon-design-system/carbon-components/blob/master/package.json#L35)

Right now, the current failing builds in Travis are because npm cannot find `@ibm/karma-ibma`...I've also excluded this [devDependency from my PR](https://github.com/carbon-design-system/carbon-components/pull/140) and Travis is passing. 

I'll keep digging to see if there are any other factors at play.

asudoh replied:
@hellobrian You are right that we have removed karma-ibma dependency from package.json, as [it has to be downloaded from https://aat.mybluemix.net/dist/karma-ibma.tgz](https://github.com/carbon-design-system/carbon-components/blob/0a9da53/.travis.yml#L42) (as it’s an internal dependency).

marijohannessen committed:
feat(date-picker): refactored the js

marijohannessen committed:
feat(date-picker): refactored the js

marijohannessen committed:
feat(date-picker): refactored the js and scss

marijohannessen committed:
feat(date-picker): added more methods

marijohannessen committed:
feat(date-picker): started writing tests

marijohannessen committed:
feat(date-picker): more test work

hellobrian commented:
A couple of NumberInput tests are failing: 

![image](https://cloud.githubusercontent.com/assets/4185382/26697666/72999f06-46d8-11e7-9cde-1ed466968b09.png)

hellobrian commented on src/components/date-picker/date-picker--range-calendar.html:
```diff
@@ -0,0 +1,18 @@
+      <label for="date-picker-2" class="bx--label">Label</label>
+      <input type="text" id="date-picker-2" class="bx--date-picker__input"  placeholder="mm / dd / yyyy" data-date-picker-input-to />
+    </div>
+      <svg data-date-picker-icon class="bx--date-picker__icon" xmlns="http://www.w3.org/2000/svg" width="17" height="19" viewBox="0 0 17 19">
```
You can remove `xmlns` attribute, it's not needed on the browsers we support.

hellobrian commented on src/components/date-picker/date-picker--single-calendar.html:
```diff
@@ -0,0 +1,14 @@
+<div class="bx--form-item">
+  <div data-date-picker="single" class="bx--date-picker">
+    <div class="bx--date-picker__container">
+      <label for="date-picker-1" class="bx--label">Label</label>
```
Even though this is for demo, make sure all associated `for` and `id` attrs are unique

tw15egan commented:
Could you remove the `yarn-error.lock` file?

hellobrian commented on src/components/date-picker/date-picker.js:
```diff
@@ -0,0 +1,207 @@
+    });
+    inputField.addEventListener('blur', () => {
+      if ((inputField.value.length === 7 || inputField.value.length === 10) && inputField.value.indexOf('/') > -1) {
+        const inputValue = inputField.value;
```
You shouldn't need to create a new variable for `inputValue`, you can just make the array variable:
```js
// const inputValue = inputField.value;
const inputValueArr = inputField.value.split('/');
```

And your `forEach` loop can use the `index` instead of the `increment` variable.
```js
inputValueArr.forEach((val, index) => {
  // increment++;
  if (index === inputValueArr.length) {
    newInputValue += `${val}`;
  } else {
    newInputValue += `${val} / `;
  }
});
```

hellobrian commented:
@marijohannessen I think you also need to work on styles for high-contrast mode. Tested in FF and I think the range selection should be made more obvious since the background styles don't really come through.

![image](https://cloud.githubusercontent.com/assets/4185382/26699945/3fac5ec2-46e1-11e7-8ba1-6c5d41948242.png)

hellobrian commented on src/components/date-picker/README.md:
```diff
@@ -0,0 +1,32 @@
+
+| Default Selector            | Description                                |
+|-----------------------------|--------------------------------------------|
+| .bx--accordion__item--active | The className for an active accordion item |
```
Don't forget to update these docs with DatePicker stuffz

marijohannessen replied:
Will do once it's ready to be merged (in case some stuff changes) :)

hellobrian commented on src/components/date-picker/_date-picker.scss:
```diff
@@ -0,0 +1,280 @@
+@import '../../globals/scss/css--reset';
+@import '../../globals/scss/typography';
+@import '../../globals/scss/import-once';
+@import 'flatpickr.scss';
```
Make sure to test that styles work when importing `_data-picker.scss` only

marijohannessen replied:
Not sure how to test this as this is the only place I'm importing `flatpickr.scss`?

hellobrian replied:
Sorry, I meant for `_date-picker.scss`. Make sure that this has all of the dependencies it needs so that users can do `@import 'date-picker'` only.

You can do test this by doing the following:
- comment out all the `@import` statements in 
- comment out the extra styles in `demo.scss`

marijohannessen replied:
Ah okay, fixed it now! Thanks :)

marijohannessen committed:
feat(date-picker): addressed comments, removed yarn file

marijohannessen commented:
@hellobrian Addressed most of your comments! Will work on the high contrast styles, thanks for catching that. No idea why the NumberInput tests are failing, I haven't touched it :-/ Will look into that as well!

asudoh commented:
Hi @marijohannessen in addition to what I mentioned earlier, I'd suggest the following changes:

1. Give the consumer of our date picker more control of Flatpickr's configuration. This is important especially for globalization perspective, given Flatpickr's configuration includes sets of localization strings. It can be done by passing along `this.options` to Flatpicker with a tweak: https://github.com/asudoh/carbon-components/commit/29b3e83ae0b7e8fda73e78df3c58ed40c7ad4088#diff-1a0dc4498ad653a61b27f80bced9f215R40
2. Separate the data attribute for date picker type from `selectorInit`, by introducing e.g. `data-date-picker-type`: https://github.com/asudoh/carbon-components/commit/29b3e83ae0b7e8fda73e78df3c58ed40c7ad4088#diff-1a0dc4498ad653a61b27f80bced9f215R28 https://github.com/asudoh/carbon-components/commit/29b3e83ae0b7e8fda73e78df3c58ed40c7ad4088#diff-1a0dc4498ad653a61b27f80bced9f215R206
  This is for allowing inherited classes to use different `selectorInit` - Otherwise only one or the other (parent class or inherited class) can be used in an application.
3. Destroying Flatpickr upon releasing our date picker: https://github.com/asudoh/carbon-components/commit/8686951a0b279f953b77929210453d944a3c4fd0#diff-1a0dc4498ad653a61b27f80bced9f215R178
  BTW I don't think the reference of Flatpickr instance has to be attached to DOM element, because application can grab our date picker instance any time via `CarbonComponents.DatePicker.components.get(element)`.

Another thing I found was that the dropdown (Flatpickr) does not seem to reflect what user directly typed in the input box - Though I haven't had time to deep give in there.

asudoh commented:
^ Another thing I'd suggest is going away from what `._addInputLogic()` does - It appears that it's for mangling date format (my apologies if I'm wrong here), but as I mentioned earlier - Different countries have different date format and the date format mangling logic looks very specific to the date format in the united states.

asudoh commented on src/globals/scss/styles.scss:
```diff
@@ -62,6 +63,7 @@ $css--reset: false !default;
 @import '../../components/progress-indicator/progress-indicator';
 @import '../../components/breadcrumb/breadcrumb';
 @import '../../components/toolbar/toolbar';
+@import '../../components/time-picker/time-picker';
```
@marijohannessen Did you really want this line? Seems that this is for time picker and I'm getting a build error complaining that this SCSS file is missing.

marijohannessen replied:
Yeah I noticed this as well, will remove it! Thanks!

asudoh commented:
Re: https://github.com/carbon-design-system/carbon-components/pull/143#issuecomment-305717466

> Another thing I found was that the dropdown (Flatpickr) does not seem to reflect what user directly typed in the input box

I found that user has to hit Enter to get the date picker updated upon input value: https://github.com/chmln/flatpickr/issues/828#issuecomment-302886760
As the issue suggests, I'd recommend hooking the `change` event to update the date picker.

marijohannessen committed:
feat(date-picker): made some more changes

marijohannessen commented:
Thanks for your comments @asudoh , will work through them today and let you know when the PR is ready for another review.

marijohannessen committed:
feat(date-picker): fixed options and config

marijohannessen committed:
feat(date-picker): removed commented code in test

marijohannessen commented:
@asudoh just updated the PR with your suggestions, let me know if it looks okay!

asudoh commented on src/components/date-picker/date-picker.js:
```diff
@@ -0,0 +1,204 @@
+    if (type === 'no-calendar') {
+      this._addInputLogic(this.element.querySelector(this.options.selectorDatePickerInput));
+    } else {
+      this.element.calendar = this._initDatePicker(type);
```
The `.release()` code refers to `this.calendar` whereas the code here refers to `this.element.calendar`. Given `DatePicker` instance associated with a DOM element can be referred to via `DatePicker.components.get(element)`, I don't think we have to attach Flatpickr instance to DOM element. Also given this I guess some of the getters/setters for interacting with Flatpickr instance (especially ones doing direct mappings) is not necessary?

marijohannessen replied:
Ah good point! I'll remove the DOM element association, and some of the getters/setters. Thanks!

asudoh commented on src/components/date-picker/date-picker.js:
```diff
@@ -0,0 +1,204 @@
+  _addInputLogic = (input) => {
+    const inputField = input;
+    inputField.addEventListener('focus', () => {
+      inputField.value = inputField.value.replace(/\s+/g, '');
```
Given we now let Flatpickr to format/parse dates - Do we still need the logic here? Note that we can use `String.prototype.trim()` for removing preceding/trailing spaces.

marijohannessen replied:
I don't think we do! It seems to be working fine without it, so I removed it. Thanks!

asudoh commented on src/components/date-picker/date-picker.js:
```diff
@@ -0,0 +1,204 @@
+    });
+    inputField.addEventListener('change', () => {
+      if (inputField.value.length === 10) {
+        const inputDate = new Date(inputField.value);
```
Letting `Date` constructor to parse date will cause globalization issues. I'd suggest using `parseDate(string, format)` API of Flatpickr to align to the parse logic Flatpickr runs internally.

marijohannessen replied:
Okay, I updated it, let me know if this is what you were thinking!

asudoh reviewed (commented):
Thanks @marijohannessen for making updates! Added some more comments for code consistency and globalization.

marijohannessen committed:
feat(date-picker): fixed demo env

marijohannessen committed:
feat(date-picker): fixed some input logic

marijohannessen committed:
feat(date-picker): temp commented out a11y test

marijohannessen committed:
feat(date-picker): added keyboard navigation

marijohannessen commented:
Updated the PR again @asudoh , let me know if you see any more issues!

marijohannessen committed:
feat(date-picker): added more imports to date picker scss

marijohannessen committed:
feat(date-picker): more work on tests

asudoh commented on src/components/date-picker/date-picker.js:
```diff
@@ -0,0 +1,222 @@
+  _addInputLogic = (input) => {
+    const inputField = input;
+    inputField.addEventListener('change', () => {
+      if (inputField.value.length === 10) {
```
Rather than checking the length which is format-specific, I’d suggest running `setDate()` *only* when `parseDate()` is successful.

marijohannessen replied:
Ah good idea, will fix! Thank you!

asudoh reviewed (commented):
Thanks alot @marijohannessen for making the changes! The code looks good now except one I put during this round of review.

marijohannessen committed:
feat(date-picker): trying to fix some issues

marijohannessen committed:
feat(date-picker): removed spacing

marijohannessen committed:
feat(date-picker): going through more tests

marijohannessen committed:
feat(date-picker): refactored a little bit and added docs

asudoh commented on src/components/date-picker/date-picker.js:
```diff
@@ -0,0 +1,210 @@
+      if (currentItem.innerHTML === 'Thu' || currentItem.innerHTML === 'Th') {
+        currentItem.innerHTML = 'Th';
+      } else {
+        currentItem.innerHTML = item.innerHTML.charAt(0);
```
@marijohannessen My apologies that not catching this earlier, but the logic trying to pick up the first character will be a problem for non-English users. The code here assumes that the days of week with 1st letter ambiguity is Tuesday and Thursday, but it'd be true only in English - For example, days of week in Chinese is "星期一", "星期二", ... (I guess they mean "first day of the week", "second day of the week", ...) where the 1st characters of them are the same for all days of week.

One option I can think of here is overriding `Flatpickr.l10n.en.weekdays.shorthand` outside `DatePicker` class definition instead.

marijohannessen replied:
Ah okay, I didn't think of that! Thanks. Will fix!

asudoh reviewed (commented):
Hi @marijohannessen thank you for making another series of changes! Found just one thing, the rest of the JavaScript code looks good. Thanks!

marijohannessen committed:
feat(date-picker): added weekday shorthand logic

marijohannessen committed:
feat(date-picker): fixed tests

asudoh commented on gulpfile.js:
```diff
@@ -284,7 +284,7 @@ gulp.task('jsdoc', cb => {
  */
 
-gulp.task('test', ['test:unit', 'test:a11y']);
+gulp.task('test', ['test:unit']);
```
Need to sort out the a11y issues the team hit. Also, if you simply disable a11y test in Travis, you can change .travis.yml instead of Gulp general test task.

asudoh reviewed (changes requested):
The JavaScript code looks good to go once we revert general Gulp test task. Thank you very much for all the efforts to get data picker in!

marijohannessen committed:
feat(date-picker): added a117 back in

marijohannessen commented:
<img width="544" alt="screen shot 2017-06-09 at 11 25 45 am" src="https://user-images.githubusercontent.com/5447411/26984791-661a1b1e-4d06-11e7-810e-e227d417638b.png">

tw15egan commented on src/components/date-picker/README.md:
```diff
@@ -0,0 +1,52 @@
+### JavaScript
+
+The date picker is built on top of [Flatpickr](https://chmln.github.io/flatpickr/), so the [events](https://chmln.github.io/flatpickr/events/) and [config options](https://chmln.github.io/flatpickr/options/) that comes with Flatpickr is therefore available to the date picker by using the reference ``. This includes methods for setting a min date, max date, disabling date(s), specifiying a range o ...
```
Is there supposed to be something inside the ``? Also, I think it should be phrased `come with Flatpickr are`

marijohannessen replied:
Ah yeah, I'll update the docs!

tw15egan commented on src/components/date-picker/_date-picker.scss:
```diff
@@ -0,0 +1,296 @@
+    padding: 0 1rem;
+    background-color: $field-01;
+    color: $text-01;
+    border: 1px solid transparent;
```
Since this seems to be re-used on all input elements, this might have enough justification to create a mixin to set the invisible borders. Also makes theming easier, as it can be set globally.

marijohannessen replied:
Good idea! Should we do that in a separate PR maybe?

tw15egan replied:
Yeah, we can just make an issue for now

marijohannessen committed:
feat(date-picker): demo changes again

tw15egan commented on src/components/date-picker/date-picker--range.html:
```diff
@@ -0,0 +1,24 @@
+<!-- Ranged -->
+<div class="bx--form-item">
+  <div data-date-picker data-date-picker-type="range" class="bx--date-picker bx--date-picker--range">
+    <div class="bx--date-picker__container">
```
For consistency, should this be `__container` or `-container`? I thought `__` was only used when we are positioning an element relatively? Correct me if I'm wrong

marijohannessen replied:
I think it's: `main-element__element` and `main-element--modifier` ? And then in the SCSS you just don't nest the BEM elements.

hellobrian replied:
FWIW, I thought we were doing it the way @tw15egan said. 
cc: @chrisdhanaraj @iangfleming

tw15egan commented on src/components/form/_form.scss:
```diff
@@ -41,15 +41,15 @@
 
     ~ .bx--form-requirement {
-      max-height: rem(200px);
+      opacity: 1;
```
Not sure if this solution will work, as this will cause inconsistent heights when using multiple form items. The text will have a height, even when hidden using this method.

marijohannessen replied:
Yeah I didn't think about that. Will fix!

tw15egan commented on src/components/form/_form.scss:
```diff
@@ -41,15 +41,15 @@
     @include font-size('11');
     margin: .75rem 0 0;
-    max-height: 0;
+    opacity: 0;
```
Same ^

tw15egan reviewed (commented):
👍 ✅ ❗️

hellobrian commented:
Merge conflict

hellobrian commented on src/components/date-picker/README.md:
```diff
@@ -0,0 +1,52 @@
+| A single date picker | [data-date-picker-type="single"] |
+| A ranged date picker | [data-date-picker-type="range"] |
+
+**The simple date picker** is a text input without a calendar. You can specify the pattern for the text input to make sure the user enters the correct date format. The default pattern that ships with the simple date picker is `\d{1,2}/\d{4}` (for short date pickers) and `\d{1,2}/\d{1,2}/\d{4}`. The simple date picker does not require any JavaScript.
```
Can you also include the date format that the regex pattern expect? 

For example, docs can be edited to say:
The default regex pattern that ships with
simple date picker is `\d{1,2}/\d{4}` ('dd/yyyy' for short date pickers) and `\d{1,2}/\d{1,2}/\d{4}` ('dd/mm/yyyy' or mm/dd/yyyy)

hellobrian reviewed (approved):
Small change to docs, then 👍 ✅ ❗️

asudoh commented on demo/views/demo-all.dust:
```diff
@@ -46,14 +46,11 @@
-  <!-- <script src="../../dist/bluemix-components.min.js"></script> -->
-  <script src="/demo.js"></script>
+  <script src="/carbon-components.min.js"></script>
+  <!-- <script src="/demo.js"></script> -->
```
@marijohannessen I quickly ran `npm run dev` using demo.js instead of carbon-components.min.js and it seems to be working. Do you still have the issues here?

hellobrian replied:
Using `<script src="/carbon-components.min.js"></script>` works for me. I didn't need to add `CarbonComponents.settings.disableAutoInit = false;` either.

marijohannessen replied:
It works when I run this locally, but it fails when I deploy it to Bluemix for visual review.. So we need to think of a process where it will work both ways.

asudoh commented on demo/views/demo-all.dust:
```diff
@@ -46,14 +46,11 @@
   <!-- true = JavaScript will not initialize automatically -->
   <!-- false = JavaScript will initialize automatically -->
-  <script>
-    CarbonComponents.settings.disableAutoInit = false;
```
Seems that this line is for [L51](https://github.com/carbon-design-system/carbon-components/pull/143/files#diff-627e32a3f2084e6bc734d3d244058336R51)-[L53](https://github.com/carbon-design-system/carbon-components/pull/143/files#diff-627e32a3f2084e6bc734d3d244058336R53) explanations, so we have two options here:

1. Removing [L51](https://github.com/carbon-design-system/carbon-components/pull/143/files#diff-627e32a3f2084e6bc734d3d244058336R51)-[L53](https://github.com/carbon-design-system/carbon-components/pull/143/files#diff-627e32a3f2084e6bc734d3d244058336R53) altogether
2. Keeping this line

marijohannessen replied:
I ended up just removing those lines, since Brian said that it worked without having to disable the auto init.

marijohannessen committed:
feat(date-picker): fixed docs and the form requirements

marijohannessen committed:
feat(date-picker): fixed demo file
